lib/auth/oidc.go

```diff
@@ -52,120 +56,103 @@ func (a *Server) getOIDCConnectorAndClient(ctx context.Context, request services
 			return nil, nil, trace.BadParameter("ConnectorID cannot be empty")
 		}
 
-		// stateless test flow
 		connector, err := types.NewOIDCConnector(request.ConnectorID, *request.ConnectorSpec)
 		if err != nil {
 			return nil, nil, trace.Wrap(err)
 		}
 
 		// we don't want to cache the client. construct it directly.
-		client, err := a.createOIDCClient(ctx, connector, false)
+		client, err := newOIDCClient(ctx, connector, request.ProxyAddress)
 		if err != nil {
 			return nil, nil, trace.Wrap(err)
 		}
-		return connector, client, nil
+		if err := client.waitFirstSync(defaults.WebHeadersTimeout); err != nil {
+			return nil, nil, trace.Wrap(err)
+		}
+
+		// close this request-scoped oidc client after 10 minutes
+		go func() {
+			ticker := a.GetClock().NewTicker(defaults.OIDCAuthRequestTTL)
+			defer ticker.Stop()
+			select {
+			case <-ticker.Chan():
+				client.syncCancel()
+			case <-client.syncCtx.Done():
+			}
+		}()
+
+		return connector, client.client, nil
 	}
 
 	// regular execution flow
 	connector, err := a.Identity.GetOIDCConnector(ctx, request.ConnectorID, true)
 	if err != nil {
 		return nil, nil, trace.Wrap(err)
 	}
-	client, err := a.getOrCreateOIDCClient(ctx, connector)
+
+	client, err := a.getCachedOIDCClient(ctx, connector, request.ProxyAddress)
 	if err != nil {
 		return nil, nil, trace.Wrap(err)
 	}
 
-	return connector, client, nil
-}
-
-func (a *Server) getOrCreateOIDCClient(ctx context.Context, conn types.OIDCConnector) (*oidc.Client, error) {
-	client, err := a.getOIDCClient(conn)
-	if err == nil {
-		return client, nil
-	}
-	if !trace.IsNotFound(err) {
-		return nil, trace.Wrap(err)
+	// Wait for the client to successfully sync after getting it from the cache.
+	// We do this after caching the client to prevent locking the server during
+	// the initial sync period.
+	if err := client.waitFirstSync(defaults.WebHeadersTimeout); err != nil {
+		return nil, nil, trace.Wrap(err)
 	}
-	return a.createOIDCClient(ctx, conn, true)
+	return connector, client.client, nil
 }
 
-func (a *Server) getOIDCClient(conn types.OIDCConnector) (*oidc.Client, error) {
+// getCachedOIDCClient gets a cached oidc client for
+// the given OIDC connector and redirectURL preference.
+func (a *Server) getCachedOIDCClient(ctx context.Context, conn types.OIDCConnector, proxyAddr string) (*oidcClient, error) {
 	a.lock.Lock()
 	defer a.lock.Unlock()
 
-	clientPack, ok := a.oidcClients[conn.GetName()]
-	if !ok {
-		return nil, trace.NotFound("connector %v is not found", conn.GetName())
-	}
+	// Each connector and proxy combination has a distinct client,
+	// so we use a composite key to capture all combinations.
+	clientMapKey := conn.GetName() + "_" + proxyAddr
 
-	config := oidcConfig(conn)
-	if ok && oidcConfigsEqual(clientPack.config, config) {
-		return clientPack.client, nil
+	cachedClient, ok := a.oidcClients[clientMapKey]
+	if ok {
+		if cmp.Equal(cachedClient.connector, conn) && cachedClient.syncCtx.Err() == nil {
```

Comment:
I see how comparing connectors instead of configs makes it possible to refactor the code more heavily, but this is will cause the clients to be recreated more often than otherwise needed. For example, any change in `ClaimToRoles` will cause a new client to be created needlessly: the config between old and new clients are the same, yet we force the rotation due to connector changes.

In practice, this is unlikely to cause problems, but I'm wary of introducing changes with the potential for subtle bugs.